lib/analysis/modules/ies_enkf.c

```diff
@@ -0,0 +1,1059 @@
+   int       subspace_dimension;      // Controlled by config key: ENKF_SUBSPACE_DIMENSION_KEY (-1: use Truncation instead)
+   double    ies_steplength;          // Step length in Gauss Newton iteration
+   double    gauss_newton_conv;       // NOT USED
+   bool      ies_subspace;            // NOT USED
+   int       ies_inversion;                // 1 for exact R, 2 for R=EE', 3 for E
+   char    * ies_logfile;             // Filename for ies logfile
+   bool      ies_debug;               // True or false for storing debug information
+   int       max_gauss_newton_it;     // NOT USED
+   int state_size;                    // Initial state_size used for checks in subsequent calls
+   bool_vector_type * ens_mask0;      // Initial ensemble mask of active realizations
+   bool_vector_type * ens_mask;       // Ensemble mask of active realizations
+   bool_vector_type * obs_mask0;      // Initial observation mask for active measurements
+   bool_vector_type * obs_mask;       // Current observation mask
+   matrix_type * W;                   // Coefficient matrix used to compute Omega = I + W (I -11'/N)/sqrt(N-1)
+   matrix_type * A0;                  // Prior ensemble used in Ei=A0 Omega_i
+   matrix_type * E;                   // Prior ensemble of measurement perturations (should be the same for all iterations)
+   bool      converged;               // GN has converged
+   ies_enkf_config_type * config;     // This I don't understand but I assume I include data from the ies_enkf_config_type defined in ies_enkf_config.c
+   bool      AAprojection;            // For including the AAprojection of Y
+};
+
+static UTIL_SAFE_CAST_FUNCTION( ies_enkf_data , IES_ENKF_TYPE_ID )
+static UTIL_SAFE_CAST_FUNCTION_CONST( ies_enkf_data , IES_ENKF_TYPE_ID )
+
+void * ies_enkf_data_alloc( rng_type * rng) {
+  ies_enkf_data_type * data = util_malloc( sizeof * data);
+  UTIL_TYPE_ID_INIT( data , IES_ENKF_TYPE_ID );
+  data->iteration_nr         = 0;
+  data->truncation           = 0.0;
+  data->subspace_dimension   = 0;
+  data->ies_steplength       = 0.0;
+  data->gauss_newton_conv    = 0.0;
+  data->ies_subspace         = false;
+  data->ies_inversion        = false;
+  data->ies_logfile          = NULL ;
+  data->ies_debug            = false;
+  data->max_gauss_newton_it  = 0;
+  data->state_size           = 0;
+  data->ens_mask0            = NULL;
+  data->ens_mask             = NULL;
+  data->obs_mask0            = NULL;
+  data->obs_mask             = NULL;
+  data->W                    = NULL;
+  data->A0                   = NULL;
+  data->E                    = NULL;
+  data->converged            = false;
+  data->config               = ies_enkf_config_alloc();
+  data->AAprojection         = true;
+  return data;
+}
+
+void ies_enkf_data_free( void * arg ) {
+  ies_enkf_data_type * data = ies_enkf_data_safe_cast( arg );
+  ies_enkf_config_free( data->config );
+  free( data );
+}
+
+void teccost(matrix_type * W,
```

Comment:
OK - there seems to be several of these `tecxxx` functions; I would recommend moving to a separate file `tecplot.cpp` or something similar.

Reply:
teccost is useful since it plots the cost function for all individual realizations as a function of the iteration step. Thus it helps to identify good realizations as well as checking that "most" realizations are reducing the cost function in each iteration step. The output format is just columns of data, but with a two-line text header on top. Maybe we could plot this in ERT instead?

tecfld is only for debugging and will be removed.

teclog is generating a log file that should be accessible to users. Again I added a header for quick plotting in tecplot, but maybe a better alternative would be to plot it directly in ERT.

Lets discuss how to best continue here.